scroll_derive/examples/derive_custom_ctx.rs

```diff
@@ -0,0 +1,72 @@
+use scroll_derive::{Pread, Pwrite, SizeWith};
+
+#[derive(Debug, PartialEq)]
+struct CustomCtx {
+    buf: Vec<u8>,
+}
+impl CustomCtx {
+    fn len() -> usize {
+        3 + 2
+    }
+}
+impl<'a> TryFromCtx<'a, usize> for CustomCtx {
+    type Error = scroll::Error;
+
+    fn try_from_ctx(from: &'a [u8], ctx: usize) -> Result<(Self, usize), Self::Error> {
+        let offset = &mut 0;
+        let buf = from.gread_with::<&[u8]>(offset, ctx)?.to_owned();
+        Ok((Self { buf }, *offset))
+    }
+}
+impl<'a> TryIntoCtx<usize> for &'a CustomCtx {
+    type Error = scroll::Error;
+    fn try_into_ctx(self, dst: &mut [u8], ctx: usize) -> Result<usize, Self::Error> {
+        let offset = &mut 0;
+        for i in 0..(ctx.min(self.buf.len())) {
+            dst.gwrite(self.buf[i], offset)?;
+        }
+        Ok(*offset)
+    }
+}
+impl SizeWith<usize> for CustomCtx {
+    fn size_with(ctx: &usize) -> usize {
+        *ctx
+    }
+}
+
+#[derive(Debug, PartialEq, Pread, Pwrite, SizeWith)]
+#[repr(C)]
+struct Data {
+    id: u32,
+    timestamp: f64,
+    #[scroll(ctx = BE)]
+    arr: [u16; 2],
+    #[scroll(ctx = CustomCtx::len())]
```

Comment:
i was surprised you could do arbitrary expressions here; maybe it's the example, but also probably because I haven't written derive code in a while, but:

1. ~I'm a bit confused what's happening here; the ctx is usize? I thought that usize as ctx's were removed in a few scroll's ago, primarily because they give illusion of being safe. so what is going on in this case then, where len() -> usize~ ?
2. not sure the exact usecase  for calling a function like that in the ctx; i'm also wondering out of pure curiosity, could you call a self method on a const or static object to also get an expr that resolves to a ctx?

Re 1., I realized as a I wrote it that the CustomCtx itself is being constructed from a usize ctx, which happens to also be the (static) method on CustomCtx::len that returns a usize.

This is actually a bit confusing and probably not the best example, I would probably change CustomCtx to CustomData, but since you implement SizeWith it is a (custom) Ctx, although it isn't used anywhere _as_ a ctx in parsing later on, so I got confused.

However, these kind of patterns are in PE binaries a lot (a thing gets parsed out using some data, and then that thing is used as the ctx later on, or a prior field is used as the length delimiter, etc.), so maybe it is a useful example? Not sure best way to make it more clear, but it's fine.

Anyway, just kind of rambling at this point :D

Reply:
I'll fix this in a bit. I agree it confusing. I was only trying to demonstrate that arbitrary expressions work, since I was also surprised that's now supported by `syn`. I think I saw somewhere that `serde` used strings when `syn` only supported strings and stays that way for compatibility. The code works by basically copy-pasting the expression entered in the various implementations. It even supports multi-line blocks (if you activate `syn`'s `full` feature) like:
```rs
{
    let a = 3;
    let b = 4;
    a + b
}
```